oeps/oep-0006.rst

```diff
@@ -0,0 +1,410 @@
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used
+    in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this
+    course, as well as settings like "allow anonymous discussion posts", "allow
+    anonymous discussion posts to peers", "discussion blackout dates", etc.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT)
+
+  * To avoid performance issues, runtimes are not required to support field
+    values larger than 4 KiB for fields that use this scope. This restriction is
+    intended to avoid the temptation to use this type of XBlock field for
+    logging or message passing. Such uses could cause data loss or performance
+    concerns, because the XBlock API does not support appends nor locking for
+    field data.
+  * Example: Enable or disable notifications about new discussion posts in this
+    course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the
+    course of the semester" (`PLAT-325`_)
+
+.. _PLAT-325: https://openedx.atlassian.net/browse/PLAT-325
+
+New XBlock view: context_settings_view
+--------------------------------------
+
+An XBlock which uses ``Scope.context_settings`` fields will need a way for
+content authors to edit the data in those fields. (e.g. when creating a course,
+the course author may wish to change course-wide settings for the discussion
+XBlock used throughout that course.)
+
+A new XBlock view, ``context_settings_view`` shall be used for this purpose.
+Just like the existing `studio_view` defined by edX Studio, this new view will
+allow an XBlock to return an arbitrary HTML Fragment which contains a form or
+other UI for editing its ``Scope.context_settings`` fields.
+
+It is expected (but not required) that runtimes like edX Studio which want to
+support editing of context-scoped fields will define a list of XBlocks that are
+enabled within a particular context/course, and will display a list of XBlocks
+that define the ``context_settings_view``. Clicking on the name of that XBlock
+would display the HTML fragment and allow the author to edit the context-scoped
+fields.
+
+.. _studio_view: http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_studio.html
+
+Context-scoped fields in OLX
+----------------------------
+
+When serializing a course with ``Scope.context_settings`` fields to Open
+Learning XML (OLX), any context-scoped fields should be described within a new
+``<xblock-settings>`` XML element that is a child of the ``<course>`` element.
+This would typically be found in ``course/course.xml``. Each XBlock class that
+has ``Scope.context_settings`` field data should be represented as a child
+element of ``<xblock-settings>``, with the name of the element matching the
+XBlock's entry point name.
```

Comment:
@ormsbee Is the following helpful? This is about the extent of my thoughts on the matter.

* Conceptually, in the OLX format, all course content and all XBlock fields are serialized as XML nodes, within a root `<course>` node [(docs)](http://edx.readthedocs.io/projects/edx-open-learning-xml/en/latest/directory-structure.html#top-level-directory). The entire course's XML may be in a single `course.xml` file or spread out into multiple files and subdirectories, where the subdirectory name is the XBlock entry point name (XML node name) and the file name is the block ID (`url_name`).
* Since context-scoped fields are XBlock fields just like Scope.content fields, they should be serialized to OLX the same way (XML, not JSON).
* Many existing course-scoped settings that are analogous to context-scoped fields (discussion settings, LTI passports, etc.) are currently stored as attributes on the `<course>` XML element.
* The option of serializing context-scoped fields as namespaced attributes on the `<course>` element is not feasible as it would be messy and conflicts with the current approach where almost every attribute is an XBlock field belonging to the `course` XBlock.
* Based on these reasons, the author recommends putting the new fields as attributes on xblock-type-specific elements that are a child of the `<course>` element. `<xblock-settings>` is suggested as a wrapper to keep these new XML nodes organized and indicate their purpose clearly.

If that is helpful I will add it to the doc.